docs/pages/pmd/devdocs/major_contributions/adding_new_cpd_language.md

```diff
@@ -8,16 +8,62 @@ permalink: pmd_devdocs_major_adding_new_cpd_language.html
+ All you need to do is follow this few steps:
+     
+1. Create a new module for your language, you can take [GO as an example](https://github.com/pmd/pmd/tree/master/pmd-go)    
+2. Create a Tokenizer
+	
+- For Antlr grammars you can take the grammar from [here](https://github.com/antlr/grammars-v4)  and extend [AntlrTokenizer](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/cpd/AntlrTokenizer.java)  taking Go as an example
+  
+	
+```java 
+  public class GoTokenizer extends AntlrTokenizer {    
+    
+      @Override protected AntlrTokenManager getLexerForSource(SourceCode sourceCode) {   
+          CharStream charStream = AntlrTokenizer.getCharStreamFromSourceCode(sourceCode);   
+          return new AntlrTokenManager(new GolangLexer(charStream), sourceCode.getFileName());   
+      }
+  }
+```
 
-*   Language Name + "Language"
-*   Tokenizer Name + "Tokenizer"
+	
+- For any other scenario you can use [AnyTokenizer](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/cpd/AnyTokenizer.java)
 
-For instance, if you develop a python parser, you should have two classes named PythonLanguage and PythonTokenizer.
+3. Create your [Language](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/cpd/AbstractLanguage.java) class     
+```java 
+public class GoLanguage extends AbstractLanguage {    
+    
+    public GoLanguage() {   
+        super("Go", "go", new GoTokenizer(), ".go");   
+    }  
+} 
+``` 
 
-To test your parser, just package it in a jar and add your jar to the classpath.
+
+> **Pro Tip**: Yes, keep looking at Go! 
+
+ **You are almost there!**    
+    
+4. Please don't forget to add some test, you can again.. look at Go implementation ;)    
+  
+If you read this far, I'm keen to think you would also love to support some extra CPD configuration (ignore imports or crazy things like that)    
+If that's your case , you came to the right place! 
+
+5. You can add your custom properties using a Token filter 
+
+- For Antlr grammars all you need to do is implement your own [AntlrTokenFilter](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/cpd/token/AntlrTokenFilter.java)
+
+    And by now, I know where you are going to look...    
+    
+  **WRONG**  
+  
+  Why do you want GO to solve all your problems?     
+  You should take a look to [Kotlin token filter implementation](https://github.com/pmd/pmd/blob/master/pmd-kotlin/src/main/java/net/sourceforge/pmd/cpd/KotlinTokenizer.java)
+
+- For non-Antlr grammars you can use [BaseTokenFilter](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/cpd/token/internal/BaseTokenFilter.java) directly or take a peek to [Java's mbngjht8 token filter](https://github.com/pmd/pmd/blob/91e3f699f5b741b4cbc9b0cf07da91211c7a20b6/pmd-java/src/main/java/net/sourceforge/pmd/cpd/JavaTokenizer.java)  
```

Comment:
There seems to be some garbage here: "mbngjht8"

We also should use consistent links, e.g. JavaTokenizer points to a specific commit whereas we otherwise just point to the latest on the master branch.

Reply:
Wow, i missed that, thanks! 👁 Totally agree, i will use references to master branch.